src/abstractdataframe/abstractdataframe.jl

```diff
@@ -2289,6 +2289,277 @@ julia> reverse(df)
    3 │     3      8     13
    4 │     2      7     12
    5 │     1      6     11
+
+julia> reverse(df, 2, 3)
+5×3 DataFrame
+ Row │ a      b      c
+     │ Int64  Int64  Int64
+─────┼─────────────────────
+   1 │     1      6     11
+   2 │     3      8     13
+   3 │     2      7     12
+   4 │     4      9     14
+   5 │     5     10     15
 ```
 """
-Base.reverse(df::AbstractDataFrame) = df[nrow(df):-1:1, :]
+Base.reverse(df::AbstractDataFrame, start::Integer=1, stop::Integer=nrow(df)) =
+    mapcols(x -> reverse(x, start, stop), df)
+
+"""
+    reverse!(df::AbstractDataFrame)
+
+Mutate data frame in-place to reverse its row order.
+
+`reverse!` will produce a correct result even if some columns of passed data frame
+are identical (checked with `===`). Otherwise, if two columns share some part of
+memory but are not identical (e.g. are different views of the same parent
+vector) then `reverse!` result might be incorrect.
+
+# Examples
+
+```jldoctest
+julia> df = DataFrame(a=1:5, b=6:10, c=11:15)
+5×3 DataFrame
+ Row │ a      b      c
+     │ Int64  Int64  Int64
+─────┼─────────────────────
+   1 │     1      6     11
+   2 │     2      7     12
+   3 │     3      8     13
+   4 │     4      9     14
+   5 │     5     10     15
+
+julia> reverse!(df)
+5×3 DataFrame
+ Row │ a      b      c
+     │ Int64  Int64  Int64
+─────┼─────────────────────
+   1 │     5     10     15
+   2 │     4      9     14
+   3 │     3      8     13
+   4 │     2      7     12
+   5 │     1      6     11
+
+julia> reverse!(df, 2, 3)
+5×3 DataFrame
+ Row │ a      b      c
+     │ Int64  Int64  Int64
+─────┼─────────────────────
+   1 │     5     10     15
+   2 │     3      8     13
+   3 │     4      9     14
+   4 │     2      7     12
+   5 │     1      6     11
+```
+"""
+function Base.reverse!(df::AbstractDataFrame, start::Integer=1, stop::Integer=nrow(df))
+    toskip = Set{Int}()
+    seen_cols = IdDict{Any, Nothing}()
+    for (i, col) in enumerate(eachcol(df))
+        if haskey(seen_cols, col)
+            push!(toskip, i)
+        else
+            seen_cols[col] = nothing
+        end
+    end
+
+    for (i, col) in enumerate(eachcol(df))
+        if !(i in toskip)
+            reverse!(col, start, stop)
+        end
+    end
+    return df
+end
+
+"""
+    permute!(df::AbstractDataFrame, p)
+
+Permute data frame `df` in-place, according to permutation `p`.
+No checking is done to verify that `p`` is a permutation.
+
+To return a new permutation, use `df[p]`.
```

Comment:
```suggestion
To return a new data frame instead of permuting `df` in-place, use `df[p]`.
```
(I know the phrasing comes from Base but it's not great IMO.)